### PR TITLE
Add child events for layout children views

### PR DESCRIPTION
**Pull request by jasonLaster**
Region events are one of the most common requets we've received:
events in a layout like you can listen for child events in a collection
view.

```js
LayoutView.extend({

  childEvents: {
    'submit': 'onChildViewSubmit'
    'fail': 'onChildViewFail'
  },

  onChildViewSubmit: function(childView, successMsg) {},
  onChildViewFail: function(childView, failMsg) {}
})
```

### Workarounds:
The first is share a radio channel,
the second is share a common eventable object (think model). Messaging
through a channel can be ugly because it encourages an explosion of
channels that can be hard to keep track of. Think lots of little wires
through the app. Messaging through a shared eventable object, is ugly
because it's a source of indirection that's unexpected. It's not obvious
that the model is being used in this way.

### Event Bubbling:
This is not event bubbling where an event will go all the way up the
view tree. I do not want to rule that out, as it might be nice, but
there are more design considerations there.

### Utility Methods
When we introduced the `_parent` reference, we discussed that there was
support for getting nested children for proxying events, but there was
no support for getting ancestor views for proxying events and other
internal utilities. We now have support for this, but will obviously
discourage end users from tapping into the power. With great power
comes...

### Docs
I'll add them when the time comes 

### Major
This should be mostly trivial to get onto major. There's one ref to LayoutView, but besides that everything should be fine

### CODE REVIEW

**Comment by jasonLaster** [comment]
After sitting on this for a day, there are two additional things that I'd like to consider:
1. adding a `childEvents` property to layoutView that works the same as collectionView
2. deleting [proxyChildEvents](https://github.com/marionettejs/backbone.marionette/blob/master/src/collection-view.js#L454) in favor of using this new approach 
3. cleaning up some variable/method names

**Comment by jasonLaster** [comment]
Just pushed updates that make use of `childEvents` and remove `proxyChildEvents` which is huge. There are some broken tests, but things are looking pretty cool. I think you guys could also help me clean up the internals a lot

**Comment by jasonLaster** [comment]
Alright, so I think this is in a good place. I went ahead with unifying implementation logic between layouts and collection views and that work is [here](https://github.com/jasonLaster/backbone.marionette/compare/marionettejs:minor...region-events2). I decided that while this works and is super cool, there is one really big drawback for unifying the. collectionView proxies `trigger` and I was only going to proxy `triggerMethod`, which is how `behaviors` work. If we want to make `CollectionView` only proxy w/r/t `triggerMethod` we *could* do it in minor I think because it passes my, it's so esoteric/rarely used nobody will notice test, but ofcourse that's more of a judgement call (semver rrrrr).

---

So, what should you look for w/ this PR:
1. layouts have a `childEvents` property
2. child view events emitted w/ `triggerMethod` will be called on the layout so `child:event` and `onChildEvent` will work.

File: src/layout-view.js [hunk 1]
```diff
@@ -14,6 +14,10 @@ Marionette.LayoutView = Marionette.ItemView.extend({
     destroyImmediate: false
   },
 
+  // used as the prefix for child view events
+  // that are forwarded through the layoutview
+  childViewEventPrefix: 'childView',
```
Comment (jasonLaster) [inline src/layout-view.js]: this will be changed to `childview` to be consistent

File: src/view.js [hunk 2]
```diff
@@ -271,15 +271,42 @@ Marionette.View = Backbone.View.extend({
...
   triggerMethod: function() {
+    var ret = Marionette._triggerMethod(this, arguments);
+
+    this._triggerEventOnBehaviors(arguments);
+    this._triggerEventOnParentLayout(arguments[0], _.rest(arguments));
+
+    return ret;
+  },
+
+  _triggerEventOnBehaviors: function(args) {
```
Comment (jasonLaster) [inline src/view.js]: small cleanup

**Comment by paulfalgout** [comment]
So I'm starting to have reservations about this.

Correctly me if I'm wrong, but this only proxies `triggerMethod` and not `trigger` ?

This is a big inconsistency with how `childEvents` works on `CollectionView`  Below is our code example from that `CollectionView` documentation which uses `trigger`  It just seems like users will expect `trigger` to work, and I don't know why we wouldn't want that too.

```
// The child view fires a custom event, `show:message`
var ChildView = new Marionette.ItemView.extend({
  events: {
    'click .button': 'showMessage'
  },

  showMessage: function () {
    console.log('The button was clicked.');

    this.trigger('show:message');
  }
});

// The parent uses childEvents to catch that custom event on the child view
var ParentView = new Marionette.CollectionView.extend({
  childView: ChildView,

  childEvents: {
    'show:message': function () {
      console.log('The show:message event bubbled up to the parent.');
    }
  }
});
```

**Comment by jasonLaster** [comment]
We could. Same mechanics. Trigger is funny cause I think onChildeventsFoo works too. I believe proxy does that for you.

File: test/unit/view.spec.js [hunk 3]
```diff
@@ -242,4 +242,74 @@ describe('base view', function() {
...
+        },
+
+        childEvents: {
+          'boom': 'onBoom'
+        },
+
+      });
+
+      this.ChildView = Marionette.ItemView.extend({
+        template: _.template('')
```
Comment (samccone) [inline test/unit/view.spec.js]: heh just use `template: false`

**Comment by jasonLaster** [comment]
NOTE: we discussed in channel adding support for `trigger` so that `this.trigger` would be handled by a layout's childEvents. We all agree that it's useful, but I'm hesitant to personally add it at this time because this PR is already complicated and I've got a lot on my plate.

---
I'll be adding some docs in a little bit after breakfast

File: docs/marionette.collectionview.md [hunk 4]
```diff
@@ -245,7 +245,7 @@ var ChildView = new Marionette.ItemView.extend({
   showMessage: function () {
     console.log('The button was clicked.');
 
-    this.trigger('show:message');
+    this.triggerMethod('show:message');
```
Comment (jasonLaster) [inline docs/marionette.collectionview.md]: we should be encouraging people to use `triggerMethod`

File: src/view.js [hunk 5]
```diff
@@ -271,15 +271,42 @@ Marionette.View = Backbone.View.extend({
...
     // Use good ol' for as this is a very hot function
     for (var i = 0, length = behaviors && behaviors.length; i < length; i++) {
-      triggerMethod(behaviors[i], arguments);
+      triggerMethod(behaviors[i], args);
     }
+  },
 
-    return ret;
+  _triggerEventOnParentLayout: function(eventName, args) {
+    var layoutView = this._parentLayoutView();
```
Comment (jasonLaster) [inline src/view.js]: I think this should be `layoutView`

File: test/unit/view.spec.js [hunk 6]
```diff
@@ -265,4 +265,74 @@ describe('base view', function() {
...
+    describe('when there is a containing layout', function() {
+      beforeEach(function(){
+        this.layoutView.showChildView('child', this.childView);
+        this.childView.triggerMethod('boom', 'foo', 'bar');
+      });
+
+      it('emits the event on the layout', function() {
+        expect(this.layoutEventHandler)
+          .to.have.been.calledWith(this.childView, 'foo', 'bar')
+          .and.CalledOnce;
```
Comment (samccone) [inline test/unit/view.spec.js]: add context and arguments plz

File: test/unit/view.spec.js [hunk 7]
```diff
@@ -265,4 +265,74 @@ describe('base view', function() {
...
+      });
+
+      it('invokes the layout on handler', function() {
+        expect(this.layoutEventOnHandler)
+          .to.have.been.calledWith(this.childView, 'foo', 'bar')
+          .and.CalledOnce;
```
Comment (samccone) [inline test/unit/view.spec.js]: add context and arguments plz

File: test/unit/view.spec.js [hunk 8]
```diff
@@ -265,4 +265,74 @@ describe('base view', function() {
...
+      });
+
+      it('invokes the layout childEvents handler', function() {
+        expect(this.layoutViewOnBoomHandler)
+          .to.have.been.calledWith(this.childView, 'foo', 'bar')
+          .and.CalledOnce;
```
Comment (samccone) [inline test/unit/view.spec.js]: add context and arguments plz

File: src/view.js (hunk 2 above)
Comment (samccone) [inline src/view.js]: can we split this change into another commit, just to keep things atomic

File: docs/marionette.view.md [hunk 9]
```diff
@@ -152,6 +152,7 @@ For more on efficient, deeply-nested view structures, refer to the LayoutView do
 This is just like the attach event described above, but it's triggered right before the view is
 attached to the document.
 
+
```
Comment (samccone) [inline docs/marionette.view.md]: :x

File: docs/marionette.collectionview.md (hunk 4 above)
Comment (samccone) [inline docs/marionette.collectionview.md]: can we split this into another commit just to be atomic

File: src/view.js [hunk 10]
```diff
@@ -271,15 +271,42 @@ Marionette.View = Backbone.View.extend({
...
+    if (!layoutView) {
+      return;
+    }
+
+    // invoke triggerMethod on parent view
+    var eventPrefix = layoutView.childViewEventPrefix;
```
Comment (samccone) [inline src/view.js]: `getChildViewEventPrefix`

File: src/view.js [hunk 11]
```diff
@@ -271,15 +271,42 @@ Marionette.View = Backbone.View.extend({
...
+    var prefixedEventName = eventPrefix + ':' + eventName;
+
+    Marionette._triggerMethod(layoutView, [prefixedEventName, this].concat(args));
```
Comment (samccone) [inline src/view.js]: so 

Just to be clear
`(eventName, ViewInstance, restOfArguments....)`

File: src/view.js [hunk 12]
```diff
@@ -271,15 +271,42 @@ Marionette.View = Backbone.View.extend({
...
+
+    // call the parent view's childEvents handler
+    var childEvents = _.result(layoutView, 'childEvents');
```
Comment (samccone) [inline src/view.js]: ya gona want to use
https://github.com/marionettejs/backbone.marionette/blob/master/src/helpers.js#L26

here

File: src/view.js (hunk 10 above)
Comment (samccone) [inline src/view.js]: https://github.com/marionettejs/backbone.marionette/blob/master/src/helpers.js#L26

Comment (samccone) [inline src/view.js]: since can be on the class, statically, a fn, or an option, static of fn

File: src/view.js (hunk 12 above)
Comment (samccone) [inline src/view.js]: since can be on the class, statically, a fn, or an option, static of fn

File: src/view.js [hunk 13]
```diff
@@ -271,15 +271,42 @@ Marionette.View = Backbone.View.extend({
...
+    var normalizedChildEvents = layoutView.normalizeMethods(childEvents);
+
+    if (!!normalizedChildEvents && _.isFunction(normalizedChildEvents[eventName])) {
```
Comment (samccone) [inline src/view.js]: ha why do we need to type case here :) either it is a hash or it is not :)

File: src/view.js [hunk 14]
```diff
@@ -300,6 +327,28 @@ Marionette.View = Backbone.View.extend({
...
 
+  // Internal utility for building an ancestor
+  // view tree list.
+  _getAncestors: function() {
+    var ancestors = [];
+    var parent  = this._parent;
+
+    while (parent) {
+      ancestors.push(parent);
+      parent = parent._parent;
```
Comment (samccone) [inline src/view.js]: Since this chain can be quite large for complex apps and this is a "hot" method, should we investigate a recursive trampolined implementation?

File: docs/marionette.layoutview.md [hunk 15]
```diff
@@ -102,6 +103,64 @@ new Marionette.LayoutView({
...
+  This also works for custom events that you might fire on your child views.
+
+  ```js
+  // The child view fires a custom event, `show:message`
+  var ChildView = new Marionette.ItemView.extend({
+    events: {
+      'click .button': 'showMessage'
+    },
+
+    showMessage: function () {
```
Comment (samccone) [inline docs/marionette.layoutview.md]: going to want to put the arguments in this.. because people like to see them

File: docs/marionette.layoutview.md [hunk 16]
```diff
@@ -102,6 +103,64 @@ new Marionette.LayoutView({
...
+      this.triggerMethod('show:message');
+    }
+  });
+
+  // The parent uses childEvents to catch that custom event on the child view
+  var ParentView = new Marionette.LayoutView.extend({
+    childView: ChildView,
+
+    childEvents: {
+      'show:message': function () {
```
Comment (samccone) [inline docs/marionette.layoutview.md]: going to want to put the arguments in this.. because people like to see them

File: docs/marionette.layoutview.md [hunk 17]
```diff
@@ -102,6 +103,64 @@ new Marionette.LayoutView({
...
+        console.log('The show:message event bubbled up to the parent.');
+      }
+    },
+
+    onChildviewShowMessage: function () {
```
Comment (samccone) [inline docs/marionette.layoutview.md]: going to want to put the arguments in this.. because people like to see them

File: docs/marionette.layoutview.md [hunk 18]
```diff
@@ -102,6 +103,64 @@ new Marionette.LayoutView({
 })
 ```
 
+### LayoutView childEvents
+
+You can specify a `childEvents` hash or method which allows you to capture all bubbling `childEvents` without having to manually set bindings. The keys of the hash can either be a function or a string that is the name of a method on the layout view.
```
Comment (samccone) [inline docs/marionette.layoutview.md]: we will want to mention what these are called with and what context

File: src/view.js (hunk 11 above)
Comment (jasonLaster) [inline src/view.js]: yup that's it

File: docs/marionette.collectionview.md (hunk 4 above)
Comment (jasonLaster) [inline docs/marionette.collectionview.md]: I think this should be grouped because it's part of the rationale.

File: src/view.js (hunk 13 above)
Comment (jasonLaster) [inline src/view.js]: meh - not sure... copy pasta

File: src/view.js (hunk 14 above)
Comment (jasonLaster) [inline src/view.js]: a wha????

File: docs/marionette.layoutview.md (hunk 15 above)
Comment (jasonLaster) [inline docs/marionette.layoutview.md]: gotta make the people happy

File: src/view.js (hunk 14 above)
Comment (samccone) [inline src/view.js]: yep it is not faster
http://jsperf.com/while-vs-recursion-mn

File: docs/marionette.collectionview.md (hunk 4 above)
Comment (samccone) [inline docs/marionette.collectionview.md]: k

File: src/view.js (hunk 14 above)
Comment (samccone) [inline src/view.js]: http://raganwald.com/2013/03/28/trampolines-in-javascript.html

File: src/view.js (hunk 2 above)
Comment (jasonLaster) [inline src/view.js]: same argument - i'd prefer to keep this refactor together.

File: src/view.js (hunk 14 above)
Comment (jasonLaster) [inline src/view.js]: i see

**Comment by samccone** [comment]
:+1: lgtm

**Comment by samccone** [comment]
@marionettejs/marionette-core can i get one more set of eyes on this... pretty sure it is good to go

**Comment by samccone** [comment]
@marionettejs/marionette-core eyes plz

File: docs/marionette.layoutview.md (hunk 15 above)
Comment (ahumphreys87) [inline docs/marionette.layoutview.md]: @jasonLaster still needs the args :)

File: docs/marionette.layoutview.md [hunk 19]
```diff
@@ -102,6 +103,77 @@ new Marionette.LayoutView({
...
+
+    showMessage: function () {
+      console.log('The button was clicked.');
+      this.triggerMethod('show:message', msg);
+    }
+  });
+
+  // The parent uses childEvents to catch that custom event on the child view
+  var ParentView = new Marionette.LayoutView.extend({
+    childView: ChildView,
```
Comment (paulfalgout) [inline docs/marionette.layoutview.md]: I don't think this is really helping?

File: test/unit/view.spec.js [hunk 20]
```diff
@@ -265,4 +265,77 @@ describe('base view', function() {
...
+      this.layoutViewOnBoomHandler = this.sinon.spy();
+      this.layoutView.onBoom = this.layoutViewOnBoomHandler;
+    });
+
+    describe('when there is not a containing layout', function() {
+      beforeEach(function(){
+        this.childView.triggerMethod('boom', 'foo', 'bar');
+      });
+
+      it('emits the event on the layout', function() {
```
Comment (paulfalgout) [inline test/unit/view.spec.js]: Should this be `it('should not emit the event on the layout, ` ?

**Comment by paulfalgout** [comment]
Couple of small notes/questions.. and I think there's still an outstanding comment about arguments in the documentation for `showMessage`

otherwise LGTM :+1:

File: docs/marionette.layoutview.md (hunk 19 above)
Comment (jasonLaster) [inline docs/marionette.layoutview.md]: nope copy pasta bug. nice catch

File: test/unit/view.spec.js (hunk 20 above)
Comment (jasonLaster) [inline test/unit/view.spec.js]: yep

File: docs/marionette.layoutview.md (hunk 15 above)
Comment (jasonLaster) [inline docs/marionette.layoutview.md]: got em

**Comment by jasonLaster** [comment]
@marionettejs/marionette-core should be all good

**Comment by paulfalgout** [comment]
:shipit: :+1:

File: src/view.js [hunk 21]
```diff
@@ -271,15 +271,42 @@ Marionette.View = Backbone.View.extend({
...
+
+    Marionette._triggerMethod(layoutView, [prefixedEventName, this].concat(args));
+
+    // call the parent view's childEvents handler
+    var childEvents = Marionette.getOption(layoutView, 'childEvents');
+    var normalizedChildEvents = layoutView.normalizeMethods(childEvents);
+
+    if (!!normalizedChildEvents && _.isFunction(normalizedChildEvents[eventName])) {
+      normalizedChildEvents[eventName].apply(layoutView, [this].concat(args));
+    }
```
Comment (jmeas) [inline src/view.js]: so much code to support this :(

**Comment by jmeas** [comment]
Lots of code to add this, but it does make Mn more consistent.

:+1:

File: src/view.js (hunk 21 above)
Comment (jasonLaster) [inline src/view.js]: Haha this is what we do. We write code. We can make it better though

**Comment by ianmstew** [comment]
:clap: Thank you, thank you, thank you for this.

File: src/view.js [hunk 22]
```diff
@@ -271,15 +271,42 @@ Marionette.View = Backbone.View.extend({
...
     var behaviors = this._behaviors;
     // Use good ol' for as this is a very hot function
     for (var i = 0, length = behaviors && behaviors.length; i < length; i++) {
-      triggerMethod(behaviors[i], arguments);
+      triggerMethod(behaviors[i], args);
     }
+  },
 
-    return ret;
+  _triggerEventOnParentLayout: function(eventName, args) {
```
Comment (megawac) [inline src/view.js]: @jasonLaster Any reason to do this on each `triggerMethod` rather than registering the events at view intitializtion and binding the functions there?

Comment (paulfalgout) [inline src/view.js]: In a future version of Mn I think I'd prefer to only use https://github.com/marionettejs/backbone.marionette/pull/2818 and remove the auto proxying of events.  we throw sooooo many events with no listeners and the auto proxying cause edge case bugs where combining collectionview proxying and layout child proxying doubles events.
